examples/lessons/12/app.js

```diff
@@ -0,0 +1,206 @@
+import {
+  GL, AnimationLoop, loadTextures, addEvents, Vector3, setParameters, Sphere, Cube
+} from 'luma.gl';
+
+import { Matrix4 } from 'math.gl';
+
+const VERTEX_SHADER = `\
+attribute vec3 positions;
+attribute vec3 normals;
+attribute vec2 texCoords;
+
+uniform mat4 uMMatrix;
+uniform mat4 uVMatrix;
+uniform mat4 uPMatrix;
+
+uniform vec3 uAmbientColor;
+
+uniform vec3 uPointLightingLocation;
+uniform vec3 uPointLightingColor;
+
+uniform bool uUseLighting;
+
+varying vec2 vTextureCoord;
+varying vec3 vLightWeighting;
+
+void main(void) {
+    vec4 mPosition = uMMatrix * vec4(positions, 1.0);
+    gl_Position = uPMatrix * uVMatrix * mPosition;
+    vTextureCoord = texCoords;
+
+    if (!uUseLighting) {
+        vLightWeighting = vec3(1.0, 1.0, 1.0);
+    } else {
+        vec3 lightDirection = normalize(uPointLightingLocation - mPosition.xyz);
+        vec4 transformedNormal = uMMatrix * vec4(normals, 0.0);
+        float pointLightWeighting = max(dot(transformedNormal.xyz, lightDirection), 0.0);
+        vLightWeighting = uAmbientColor + uPointLightingColor * pointLightWeighting;
+    }
+}
+`;
+
+const FRAGMENT_SHADER = `\
+#ifdef GL_ES
+precision highp float;
+#endif
+
+varying vec2 vTextureCoord;
+varying vec3 vLightWeighting;
+
+uniform sampler2D uSampler;
+
+void main(void) {
+  vec4 textureColor = texture2D(uSampler, vec2(1.0 - vTextureCoord.s, 1.0 - vTextureCoord.t));
+  gl_FragColor = vec4(textureColor.rgb * vLightWeighting, textureColor.a);
+}
+`;
+
+const moonRotation = {
+  matrix: new Matrix4().rotateY(degToRad(180)).translate([5, 0, 0])
+};
+
+const cubeRotation = {
+  matrix: new Matrix4().translate([5, 0, 0])
+};
+
+const animationLoop = new AnimationLoop({
+  onInitialize: ({canvas, gl}) => {
+
+    setParameters(gl, {
+      clearColor: [0, 0, 0, 1],
+      clearDepth: 1,
+      depthTest: true
+    });
+
+    return loadTextures(gl, {
+      urls: ['moon.gif']
+    })
+    .then(textures => {
+      let moon = new Sphere(gl, {
+        fs: FRAGMENT_SHADER,
+        vs: VERTEX_SHADER,
+        uniforms: {
+          uSampler: textures[0]
+        },
+        nlat: 30,
+        nlong: 30,
+        radius: 2,
+      });
+      return loadTextures(gl, {
+        urls: ['crate.gif']
+      })
+      .then(textures => {
+        let cube = new Cube(gl, {
+          vs: VERTEX_SHADER,
+          fs: FRAGMENT_SHADER,
+          uniforms: {
+            uSampler: textures[0]
+          }
+        });
+        return {moon, cube};
+      });
+    });
+  },
+  onRender: ({
+    gl, tick, aspect, moon, cube
+  }) => {
+    // Update Camera Position
+    const eyePos = [0, 0, 20];
+
+    gl.clear(GL.COLOR_BUFFER_BIT | GL.DEPTH_BUFFER_BIT);
+
+    let uMoonMMatrix = new Matrix4()
```

Comment:
Doesn't this just set `uMoonMatrix` to `moonRotation.matrix`?

Can't just pass `moonRotation.matrix` to `uMoonMatrix` below in set uniforms?

Reply:
Good catch, thanks! 👍